### ISSUE TITLE
Allow user defined DevServices in Dev mode

### ISSUE THREAD

**Issue by carlosthe19916** [quarkusio/quarkus#21196]
### Description

Currently, there are a predefined set of DevServices available in Quarkus like Databases, Kafka, Keycloak, etc. The list of DevServices is listed in the official documentation https://quarkus.io/guides/dev-services . What makes DevServices powerful is that it allows developers to start the project without being worried about dependencies like databases, Kafka, Keycloak, etc. So a developer can easily execute `./mvnw compile quarkus:dev` and the magic happens, hurray!

Although the predefined set of DevServices helps a lot, it is not enough since there is an unlimited number of dependencies an application might depend on; for instance, if an application depends on Elasticseach then the developer will need to start that dependency by himself since it is not provided by Quarkus. Like Elasticseach there are different dependencies and Quarkus can not create a DevService for each dependency that exists out there.

Currently, if there is a dependency not provided by DevServices then it is possible to create a class implementing the interface `QuarkusTestResourceLifecycleManager` and using TestContainers to start a new container with all the configurations we want to have. Here is an example of how a developer can have its own dependency started:

```java
public class ElasticsearchServer implements QuarkusTestResourceLifecycleManager {

    private GenericContainer<?> elasticsearch;

    @Override
    public Map<String, String> start() {
        elasticsearch = new GenericContainer<>("docker.elastic.co/elasticsearch/elasticsearch:" + System.getProperty("elasticsearch.version", "7.10.2"))
                .withExposedPorts(9200, 9300)
                .withEnv("discovery.type", "single-node")
                .waitingFor(
                        new HttpWaitStrategy()
                                .forPort(9200)
                                .forStatusCodeMatching(response -> response == HTTP_OK || response == HTTP_UNAUTHORIZED)
                                .withStartupTimeout(Duration.ofMinutes(2))
                );

        elasticsearch.start();

        String host = elasticsearch.getHost();
        Integer port = elasticsearch.getMappedPort(9200);

        return new HashMap<>() {{
            put("quarkus.hibernate-search-orm.elasticsearch.hosts", host + ":" + port);
            put("quarkus.hibernate-search-orm.elasticsearch.version", "7");
        }};
    }

    @Override
    public void stop() {
        elasticsearch.stop();
    }
}
```

# If we have already QuarkusTestResourceLifecycleManager then what is the problem?
The big problem is that `QuarkusTestResourceLifecycleManager` only works while executing tests but it doesn't help developers to start the app with a single command `./mvnw compile quarkus:dev`. What I mean, is that the TEST experience is covered by `QuarkusTestResourceLifecycleManager` but the `./mvnw compile quarkus:dev` is not.

It is just an idea: would it be possible to somehow do something so developers can define custom dependencies (using TestsContainers) that will later be used in Dev and also in Test?

### Implementation ideas

Probably a stupid idea:
Could we move classes that inherit `QuarkusTestResourceLifecycleManager` to the `src/main/java` folder and execute those classess also when developers execute `./mvnw compile quarkus:dev`? Of course, when we build the application all classes that inherit `QuarkusTestResourceLifecycleManager` must be omitted since they are not really part of the app but just helpers.

Please let me know if this is a crazy idea that should be thrown to the garbage or we might do something about it in the future depending on the priorities. I do believe the Developer's experience would be enhanced if such a thing is possible to do.

**Comment by stuartwdouglas** [comment]
I had actually made a start on something like this here: https://github.com/stuartwdouglas/quarkus/tree/generic-devservices

Basically it allows for you to specify image names to run as custom dev services. These services can also interact with other services (e.g. connect to the same database/kafka).

There did not seem to be a great deal of interest when I posted on the mailing list, so I abandoned it. That said maybe something programmatic like you suggest could be potentially more useful and easier to use.
